### PR TITLE
Adds column preselection for readtable ('usecols')

### PR DESCRIPTION
**Pull request by lettis**
Hi,
I recently implemented a 'usecols' option for readtable.
It is designed analogously to the according read_csv-option in pandas.
It has especially high value when working with huge data files, whose complete parsing would fill the entire RAM...

'usecols' expects one of two kinds of values: either Vector{Int} with column indices (e.g. [1,3,4] for first, third and fourth columns), or Vector{Symbol} with column names.
Column names are previously parse, thus one can either use the names as defined in a header, manually defined names, or automatically defined symbols (e.g. :x1, :x2, ...).

If an attempt is done to read a non-existing column, an appropriate error is thrown.

The commits include the updated source, documentation and tests.

Fixes issues #605 and #568.

Open question: There is a 'filldf!' function, which looks awfully similar to 'builddf'.
However, it does not seem to be used anywhere and is not exported. Thus, I have not implemented 'usecols' for it. Was that a misjudgement? It would be easy for me, to adapt 'filldf!', too.
Feedback welcome.

Cheers,
  Florian

### CODE REVIEW

File: src/dataframe/io.jl [hunk 1]
```diff
@@ -505,9 +506,33 @@ function builddf(rows::Integer,
...
+    if isempty(o.usecols)
+        colindices = 1:cols
+    else
+        if isa(o.usecols, Vector{Int})
+            colindices = indexin(o.usecols, 1:cols)
+        else
+            colindices = indexin(o.usecols, names)
+        end
+        if any(colindices .== 0)
+            error("Invalid column selection")
```
Comment (nalimilan) [inline src/dataframe/io.jl]: Could you make the error message more detailed? I know it's terribly frustrating to get this general error instead of the name or index of the incorrect column identifier. Mentioning `usecols` would also be helpful.

File: src/dataframe/io.jl [hunk 2]
```diff
@@ -946,6 +970,8 @@ function filldf!(df::DataFrame,
         end
     end
 
+    #TODO implement 'usecols'-optoins
```
Comment (nalimilan) [inline src/dataframe/io.jl]: `filldf!` was added by @johnmyleswhite in a commit (https://github.com/JuliaStats/DataFrames.jl/commit/c7b1f315a89118bb1385758b1a0781e131937e32) that didn't use nor export it. John, do you think this code is still useful?

File: test/io.jl [hunk 3]
```diff
@@ -282,6 +282,25 @@ module TestIO
     @test df[:c5][2] == "true"
     @test df[:c5][3] == "true"
 
+    # Readtable column preselection
+    filename = "$data/typeinference/standardtypes.csv"
+    df = readtable(filename)
+    @test size(df) == (3,5)
+    df = readtable(filename, usecols=[1,3,5])
```
Comment (nalimilan) [inline test/io.jl]: A line break between each series of test wouldn't hurt. Also, you could mention `usecols` in the comment above. Finally, could you check for corner cases like providing duplicate column identifiers (should it work?), and invalid ones (use `@test_throws` to check that it fails)?

**Comment by nalimilan** [comment]
Thanks, this sounds useful. The tests failures on 0.3 and 0.5 do not seem related (though they are annoying and should be fixed elsewhere).

Please see https://github.com/JuliaStats/DataFrames.jl/pull/910#issuecomment-182824200 about documenting the new argument.

**Comment by lettis** [comment]
No problem. I will tend to the remarks tonight and send an update.
Indeed, the test failures were already in the master branch when I pulled.

**Comment by nalimilan** [comment]
See https://github.com/JuliaStats/DataFrames.jl/pull/912 about fixing tests.

File: src/dataframe/io.jl (hunk 1 above)
Comment (lettis) [inline src/dataframe/io.jl]: done

File: test/io.jl (hunk 3 above)
Comment (lettis) [inline test/io.jl]: done.
columns are now sanitized (no duplicates and in order of original columns).
this also mimics the pandas-behavior.

missing tests added.

**Comment by lettis** [comment]
Added missing doc in new markdown docs, added tests, implemented column selection sanitizing (remove duplicates, sort indices).

@nalimilan 
I would leave the test errors to #912, since it's a pull request on its own and may be merged into master by its own right.
If you'd rather me to pull the changes into this repo, please tell me.

**Comment by nalimilan** [comment]
Thanks, looks good to me. I'll let others comment, but if there are no objections you could squash the commits into one to be ready to merge.

**Comment by lettis** [comment]
Hi, I have squashed the commits into one. However, it was necessary to pull from my github-branch and merge in a second commit. I do not really understand why, since I hadn't changed anything on the github-branch. Is that normal behavior, or did I do something wrong?

Cheers,
  Florian

**Comment by nalimilan** [comment]
I guess the problem is that you have to force-push to overwrite commit history. You can do that with `git push lettis +lts/usecols`. (But before that, squash again, including the merge commit.)

**Comment by lettis** [comment]
ok, done. thanks for the hint and sorry for the mess.
everything is now squashed together.

**Comment by nalimilan** [comment]
I've just merged test fixes to master. If you can rebase on that, it should show that the tests pass on  your branch.

**Comment by lettis** [comment]
I've just pulled from master and merged it with my branch.
Unfortunately, it does not seem to have improved the situation...

**Comment by nalimilan** [comment]
Cool. So in the end it was a good idea to really check the tests.

Could anybody else comment before merging?

**Comment by alyst** [comment]
I hadn't looked in the code thoroughly, but how the `eltypes` option is handled by this PR? Would `eltypes` only refer to the columns selected or the original ones?

I think `usecols` is a useful addition, but maybe it's possible to have something more powerful that would also combine `eltypes` and allow specifying columns to exclude, instead of the ones to include?
I'm thinking about something like `colinfo::Dict{UTF8String, DataType}`, which assigns column type by its name. E.g. `colinfo["X"] = Any` would specify that we need to read column "X", the type would be automatically determined by `readtable()`, `colinfo["X"] = Void` would mean that "X" have to be excluded. There should be also the key for the default behaviour, e.g. `colinfo["__DEFAULT__"] = Void` instructs to skip any column that is not explicitly listed in `colinfo`. Of course, `usecols` and `eltypes` options could be also available as simpler alternatives that are internally converted to `colinfo`. This would help streamlining the column handling code inside `readtable()`.

**Comment by nalimilan** [comment]
Sounds interesting. It should also allow passing a simple vector of types to specify the column types without their names (i.e. in order of appearance), and a vector of symbols/integers to give a list of columns to retain when you don't want to repeat `Any` for each of them.

**Comment by lettis** [comment]
`eltypes` is indeed currently unchecked against `usecols`, which needs to be changed.

Personally, I would opt for using both options, one for selecting the columns and one for defining their data type. A single `colinfo` option that may select and/or exclude certain columns and, at the same time, define their data type seems to me overly complicated from a user view.

Personally, I do not have the use case of loading all but a few columns. I do not know how interesting column exclusion would be for others.

**Comment by alyst** [comment]
@lettis Some tables I'm working with could have hundreds of columns ("wide" format). In such scenario it's more straightforward to skip a few selected ones and specify the data type for a few others, where `readtable()` makes a wrong guess. Rather then enumerating all columns to include.

**Comment by nalimilan** [comment]
Another, simpler solution is to use a `Not` type like NamedArrays: https://github.com/davidavdav/NamedArrays.jl/blob/5310388fafc069639ce4404b2146054e129ad254/src/namedarraytypes.jl#L29 Passing `Not(:colA)` or `Not(1)` would mean that all coumns must be retained except this one. This pattern might enter Base at some point, since there is currently no way to do this (which e.g. R supports with negative indices).

Independently from this, we could allow passing a `(column, type)` dictionary as `eltypes`, giving the types of the specified columns. That sounds simpler to grasp and to implement to me than a common argument for `usecols` and `eltypes`.

**Comment by lettis** [comment]
@alyst Indeed, that is a good use case for exlusion of colums ;)

@nalimilan I really like the `Not(...)`-idea.
However, I would define it as `Not([...])`, to keep the structure of vectors of symbols or integers, as used for the selection.
For `eltypes`, the type would then be a Union of a type-vector (like it is done now), and the new dictionary?
I think that is a good solution. It keeps column selection and data type definition apart (and thus easy to grasp).

**Comment by mirestrepo** [comment]
Is there any news on this? Having usecols options will indeed be very useful

**Comment by lettis** [comment]
Hi,
it's on my ToDo-list, but at the moment I unfortunately do not find the time for it.
I am in the last months of my PhD and will probably be able to get back to this after handing in, which is in about half a year.
However, I'll maybe find some spare time in between...

Cheers,
  Florian

**Comment by quinnj** [comment]
`readtable` `writetable` are now deprecated.
